### PR TITLE
Incremental stat var hierarchy

### PR DESCRIPTION
**Pull request by shifucun**
* Load the children stat var and stat var group nodes on click; Remove the preload full hierarchy on page load.
* Use stat var search API in mixer and remove all the stat var hierarchy search code.
* Simplify various components props and states

### CODE REVIEW

File: static/package-lock.json [hunk 1]
```diff
@@ -8596,7 +8596,10 @@
       "version": "1.3.1",
       "resolved": "https://registry.npmjs.org/full-icu/-/full-icu-1.3.1.tgz",
       "integrity": "sha512-VMtK//85QJomhk3cXOCksNwOYaw1KWnYTS37GYGgyf7A3ajdBoPGhaJuJWAH2S2kq8GZeXkdKn+3Mfmgy11cVw==",
-      "dev": true
+      "dev": true,
```
Comment (chejennifer) [inline static/package-lock.json:8599]: Did you mean to include this file in this pr as well?

File: static/js/tools/map/context.ts [hunk 2]
```diff
@@ -58,7 +58,7 @@ export interface StatVarInfoWrapper {
 // Information relating to the places to plot
 export interface PlaceInfo {
   // Place that encloses the places to plot
-  enclosingPlace: NamedPlace;
+  enclosingPlace: Node;
```
Comment (chejennifer) [inline static/js/tools/map/context.ts:61]: Just wondering, why the change from NamedPlace to Node? I feel like Node is a very generic name that could be confusing

**Review by chejennifer — COMMENTED** [review]
nice changes! Just a few comments that need updating

File: static/js/browser/statvar_group_node.tsx [hunk 3]
```diff
@@ -36,16 +41,12 @@ interface StatVarGroupNodePropType {
   // The path of this stat var group node.
   path: string[];
   // A list of named places object.
-  places: NamedPlace[];
-  // the dcid of the current stat var group
-  statVarGroupId: string;
+  places: Node[];
   // the mapping of all stat var groups to their corresponding information
-  data: { [key: string]: StatVarGroupNodeType };
+  data: StatVarGroupInfo;
```
[the remark this answers is not in the document]
Comment (shifucun) [reply static/js/browser/statvar_group_node.tsx:46]: Done

File: static/js/tools/map/context.ts (hunk 2 above)
Comment (shifucun) [reply static/js/tools/map/context.ts]: was trying to unify the type, but you are right, NamedPlace is better here. I added type alias with NamedNode = NamedPlace now and reverted these changes

File: static/js/browser/statvar_hierarchy.tsx [hunk 4]
```diff
@@ -54,6 +54,8 @@ interface StatVarHierarchyStateType {
   errorMessage: string;
   // Indicates when search input was cleared after a result was selected.
   searchSelectionCleared: boolean;
+  // A map from stat var group dcid to the metadata.
```
[the remark this answers is not in the document]
Comment (shifucun) [reply static/js/browser/statvar_hierarchy.tsx]: Done

File: server/routes/api/browser.py [hunk 5]
```diff
@@ -197,11 +195,31 @@ def get_statvar_hierarchy_helper(dcid_string):
...
+    stat_var_group = request.args.get("stat_var_group")
+    places = request.args.getlist("places")
+    result = dc.get_statvar_group(stat_var_group, places)
+    return Response(json.dumps(result), 200, mimetype='application/json')
+
+
+@bp.route('/statvar/path')
+@cache.cached(timeout=3600 * 24, query_string=True)
+def get_statvar_path():
+    """Gets the statvars and statvar groups that match the tokens in the query
```
[the remark this answers is not in the document]
Comment (shifucun) [reply server/routes/api/browser.py]: Done

File: server/routes/api/browser.py [hunk 6]
```diff
@@ -197,11 +195,31 @@ def get_statvar_hierarchy_helper(dcid_string):
...
+    query = request.args.get("query")
+    places = request.args.getlist("places")
+    result = dc.search_statvar(query, places)
+    return Response(json.dumps(result), 200, mimetype='application/json')
+
+
+@bp.route('/statvar/group')
+@cache.cached(timeout=3600 * 24, query_string=True)
+def get_statvar_group():
+    """Gets the statvars and statvar groups that match the tokens in the query
```
[the remark this answers is not in the document]
Comment (shifucun) [reply server/routes/api/browser.py]: Done

File: static/package-lock.json (hunk 1 above)
Comment (shifucun) [reply static/package-lock.json:8599]: reverted

**Review by chejennifer — APPROVED** [review]
Thanks for updating

**Review by beets — APPROVED** [review]
nice, the singleton-node expansion is very cool.

File: static/js/browser/statvar_group_node.tsx [hunk 7]
```diff
@@ -150,6 +175,35 @@ export class StatVarGroupNode extends React.Component<
     );
   }
 
+  private fetchData(): void {
+    if (this.state.dataFetched) {
+      return;
+    }
+    // stat var (group) dcid count contain "&", need to encode here.
```
[the remark this answers is not in the document]
Comment (pradh) [reply static/js/browser/statvar_group_node.tsx:182]: Not a documented list :(.  The special chars are generally:  /_-.&

It is interesting why the this & comes about.  Happens in cases where we effectively concatenate two enums in a single V value. That is because P values in SV cannot be repeated, and we don't want to define new enums just to represent combinations.

[the remark this answers is not in the document]
Comment (shifucun) [reply static/js/browser/statvar_group_node.tsx]: Thanks for the explanation, the example is like "dc/g/Person_Citizenship-NotAUSCitizen_CorrectionalFacilityOperator-StateOperated&FederallyOperated&PrivatelyOperated"

File: static/js/browser/statvar_group_node.tsx [hunk 8]
```diff
@@ -57,6 +58,10 @@ interface StatVarGroupNodeStateType {
   // open, we want to render an expanded collapsible by passing in true for the
   // open prop.
   toggledOpen: boolean;
+  childSVG: StatVarGroupInfo[];
```
[the remark this answers is not in the document]
Comment (shifucun) [reply static/js/browser/statvar_group_node.tsx:62]: Done

File: server/webdriver_tests/ranking_test.py [hunk 9]
```diff
@@ -54,7 +54,7 @@ def test_northamerica_population(self):
         x_text = chart.find_elements_by_class_name(
             'x')[0].find_elements_by_tag_name('text')
         self.assertEqual(x_text[0].text, 'United States of America')
-        self.assertEqual(x_text[-1].text, 'Saint Pierre and Miquelon')
+        self.assertEqual(x_text[-1].text, 'Saint Barthélemy')
```
[the remark this answers is not in the document]
Comment (shifucun) [reply server/webdriver_tests/ranking_test.py]: Reverted.

File: static/js/browser/statvar_group_node.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (beets) [reply static/js/browser/statvar_group_node.tsx]: If the `&` is a new addition with SVG, we'll need to test it with our other pages, eg browser. We have not had to worry about URL search special chars before, so the encoding might need to be propagated.

Could we use `_and_` or `,` instead?

[the remark this answers is not in the document]
Comment (shifucun) [reply static/js/browser/statvar_group_node.tsx]: That's a good point, any page with this stat var will not show properly..

[the remark this answers is not in the document]
Comment (pradh) [reply static/js/browser/statvar_group_node.tsx]: Good points.

1. At the moment, there are only 2 SV IDs with '&' in it (https://screenshot.googleplex.com/83cTHghienxMjUS).
2. That SV node [shows up](https://datacommons.org/browser/USStateQuarterlyIndustryGDP_NAICS_311_316&322_326), but naics PV is missing looks like. Also, oddly there are no SVObs for these SVs (https://screenshot.googleplex.com/6bySnwSjjfvaHuD). So I think nothing rn is too broken.
3. And yes, we can remove '&' from the SVG ID.

`,` is a delimiter too. but `_And_` works.
